### PR TITLE
[sdpt-526] support CSV and other types

### PR DESCRIPTION
**Pull request by chennv4**
**Problem:**
Existing REST API supports only JSON. We need to support CSV also. Create a common API to support both.

**Implementation:**
Added a API
    @POST
    @Path("/rawData/{remoteAddr}")
    public String postRawData(@Context Request request, byte[] data, @PathParam("remoteAddr") String remoteAddr)

**Test Results**

DEBUG [2020-06-21 03:40:44.728] [grizzly-http-server-0] i.p.example.iot.gateway.DataHandler: routingKey=default_routingKey, message=
"message_id","timestamp","device_id","temperature","humidity","RemoteAddr","IngestTimestamp","IngestTimestampStr"
"111111","1588203203770","098765","60.96","80.4","my_routing_key_2","1592613114765","2020-06-20T10:31:54.763Z"
"222222","1588203203770","098765","60.96","80.4","my_routing_key_2","1592613114765","2020-06-20T10:31:54.763Z"
"333333","1588203203770","098765","60.96","80.4","my_routing_key_2","1592613114765","2020-06-20T10:31:54.763Z"
"444444","1588203203770","098765","60.96","80.4","my_routing_key_2","1592613114765","2020-06-20T10:31:54.763Z"
"555555","1588203203770","098765","60.96","80.4","my_routing_key_2","1592613114765","2020-06-20T10:31:54.763Z"

### CODE REVIEW

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java [hunk 1]
```diff
@@ -68,7 +70,15 @@ public String postData(@Context Request request, String data, @PathParam("remote
 
                 // Write the message to Pravega.
                 Log.debug("routingKey={}, message={}", routingKey, message);
-                final CompletableFuture<Void> writeFuture = Main.getWriter().writeEvent(routingKey, message);
+                // Writing...
+                // Create an immutable writer (in this case using the default settings)
+                final ObjectWriter writer = objectMapper.writer();
```
Comment (claudiofahey) [inline gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: There is no need to create an ObjectWriter. Just call objectMapper.writeValueAsBytes().

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java [hunk 2]
```diff
@@ -83,4 +93,41 @@ public String postData(@Context Request request, String data, @PathParam("remote
             throw e;
         }
     }
+
+    @POST
+    @Path("/rawData/{remoteAddr}")
```
Comment (claudiofahey) [inline gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: Add `@Produces({"application/json"})`

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java [hunk 3]
```diff
@@ -83,4 +93,41 @@ public String postData(@Context Request request, String data, @PathParam("remote
...
+    public String postRawData(@Context Request request, byte[] data, @PathParam("remoteAddr") String remoteAddr) throws Exception {
+        final long ingestTimestamp = System.currentTimeMillis();
+        final String ingestTimestampStr = dateFormat.format(new Date(ingestTimestamp));
+        try {
+            if(data == null)
+                return "{DATA Not Received}";
```
Comment (claudiofahey) [inline gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: Will this method ever receive a null value for data? If the HTTP body is empty, won't this send an empty byte[]? If a null value is not a possible input, then there is no reason to test for it. Even if it does get a null, this method will just throw a NPE.

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java (hunk 2 above)
Comment (claudiofahey) [inline gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: Rename remoteAddr parameter to routingKey.

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java [hunk 4]
```diff
@@ -83,4 +93,41 @@ public String postData(@Context Request request, String data, @PathParam("remote
...
+
+            // Get or calculate the routing key.
+            final String routingKeyAttributeName = Parameters.getRoutingKeyAttributeName();
```
Comment (claudiofahey) [inline gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: Simplify this section to use the routingKey value from the URL path. Do not use Parameters.getRoutingKeyAttributeName() and do not use Math.random().

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java [hunk 5]
```diff
@@ -83,4 +93,41 @@ public String postData(@Context Request request, String data, @PathParam("remote
...
+            final String routingKey;
+            if (routingKeyAttributeName.isEmpty()) {
+                routingKey = Double.toString(Math.random());
+            }else {
+                routingKey = "default_routingKey";
+            }
+
+            String message = new String(data);
```
Comment (claudiofahey) [inline gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: Are you guaranteed that new `String(data)` will always succeed? What if data contains binary data such as an Avro event? I recommend not trying to parse data in any way. Change your logging method to only log the size of the data, not the contents.

File: scripts/send-test-event.sh [hunk 6]
```diff
@@ -6,15 +6,14 @@ source ${ROOT_DIR}/scripts/env-local.sh
 export CURL_CA_BUNDLE="${ROOT_DIR}/ssl/data/ca.pem"
 
 curl -v \
---header "Content-Type: application/json" \
+--header "Content-Type: application/csv" \
```
Comment (claudiofahey) [inline scripts/send-test-event.sh]: Create two test scripts. One to write a JSON event (like the original) and another to send a CSV.

**Review by claudiofahey — CHANGES_REQUESTED** [review]

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java (hunk 2 above)
Comment (chennv4) [reply gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: added

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java (hunk 1 above)
Comment (chennv4) [reply gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: modified accordingly

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java (hunk 3 above)
Comment (chennv4) [reply gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: good point. removed the check.

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java (hunk 4 above)
Comment (chennv4) [reply gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: removed

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java (hunk 2 above)
Comment (chennv4) [reply gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: changed

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java (hunk 5 above)
Comment (chennv4) [reply gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java]: changed to print length of the message.

File: scripts/send-test-event.sh (hunk 6 above)
Comment (chennv4) [reply scripts/send-test-event.sh]: created a new script for csv

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java [hunk 7]
```diff
@@ -83,4 +86,33 @@ public String postData(@Context Request request, String data, @PathParam("remote
             throw e;
         }
     }
+
+    @POST
+    @Path("/rawData/{routingKey}")
+    @Produces({"application/json"})
+    public String postRawData(@Context Request request, byte[] data, @PathParam("routingKey") String routingKey) throws Exception {
+        final long ingestTimestamp = System.currentTimeMillis();
```
Comment (claudiofahey) [inline gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java:94]: This is not used.

File: gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java [hunk 8]
```diff
@@ -83,4 +86,33 @@ public String postData(@Context Request request, String data, @PathParam("remote
...
+        final String ingestTimestampStr = dateFormat.format(new Date(ingestTimestamp));
+        try {
+            if (routingKey.isEmpty()) {
```
Comment (claudiofahey) [inline gateway/src/main/java/io/pravega/example/iot/gateway/DataHandler.java:97]: This is not needed. The routing key can be an empty string.

**Review by claudiofahey — APPROVED** [review]
Approved. I made some minor updates and it is now merged.

**Pull request merged**